bluemira/equilibria/flux_surfaces.py

```diff
@@ -872,3 +876,251 @@ def poloidal_angle(Bp_strike, Bt_strike, gamma):
+                    ( P_l * cos(theta_f) / sqrt(l*(l+1)) )
+
+    Where l = degree, and P_l * cos(theta_f) are the associated
+    Legendre polynomials of degree l and order (m) = 1.
+
+    Parmeters
+    ----------
+    x_f: np.array
+        X coordinates of filaments (coils)
+    z_f: np.array
+        Z coordinates of filaments (coils)
+    i_f: np.array
+        Currents of filaments (coils)
+    max_degree: integer
+        Maximum degree of harmonic to calculate up to
+    r_t: float
+        Typical length scale (e.g. radius at outer midplane)
+
+    Returns
+    -------
+    amplitudes: np.array
+        Array of spherical harmonic amplitudes from given coil potitions and currents
+     max_valid_r: float
+        Maximum spherical radius for which the spherical harmonics apply
+    """
+    # SH coeffcients from fuction of the current distribution outside of the sphere
+    # containing the plamsa, i.e., LCFS (r_lcfs)
+    # SH coeffs = currents2harmonics @ coil currents
+    # N.B., max_valid_r >= r_lcfs,
+    # i.e., cannot use coil located within r_lcfs as part of this method.
+    currents2harmonics, max_valid_r = coil_harmonic_amplitude_matrix(
+        x_f, z_f, max_degree, r_t
+    )
+
+    return currents2harmonics @ i_f, max_valid_r
+
+
+def coil_harmonic_amplitude_matrix(x_f, z_f, max_degree, r_t):
+    """
+    Construct matrix from harmonic amplitudes at given coil locations.
+
+    Parmeters
+    ----------
+    x_f: np.array
+        X coordinates of filaments (coils)
+    z_f: np.array
+        Z coordinates of filaments (coils)
+    max_degree: integer
+        Maximum degree of harmonic to calculate up to
+    r_t: float
+        Typical length scale (e.g. radius at outer midplane)
+
+    Returns
+    -------
+    currents2harmonics: np.array
+        Matrix of harmonic amplitudes (to get spherical harmonic coefficents
+        -> matrix @ vector of coil currents, see coil_harmonic_amplitudes)
+     max_valid_r: float
+        Maximum spherical radius for which the spherical harmonic
+        approximation is valid
+    """
+    # Spherical coords
+    r_f = np.sqrt(x_f**2 + z_f**2)
+    theta_f = np.arctan2(x_f, z_f)
+    # Maxmimum r value for the sphere whithin which harmonics apply
+    max_valid_r = np.amin(r_f)
+
+    # [number of degrees, number of coils]
+    currents2harmonics = np.zeros([max_degree, np.size(r_f)])
+    # First 'harmonic' is constant (this line avoids Nan isuues)
+    currents2harmonics[0, :] = 1  #
+
+    # SH coeffcients from fuction of the current distribution
+    # outside of the sphere coitaining the LCFS
+    # SH coeffs = currents2harmonics @ coil currents
+    for degree in np.arange(1, max_degree):
+        currents2harmonics[degree, :] = (
+            0.5
+            * MU_0
+            * (r_t / r_f) ** degree
+            * np.sin(theta_f)
+            * lpmv(1, degree, np.cos(theta_f))
+            / np.sqrt(degree * (degree + 1))
+        )
```

Comment:
A lot of this is repeated below, no? This looks like it could be something like:

```python
currents2harmonics = 0.5 * MU_0 * harmonic_amplitude_matrix(r_t/r_f, theta_f, max_degree, 1)
```

It looks like a transpose might be necessary though, seeing as you have:

```python
    # [number of degrees, number of coils]
    currents2harmonics = np.zeros([max_degree, np.size(r_f)])

    # [number of points, number of degrees]
    harmonics2collocation = np.zeros([n_collocation, max_degree])
```

I have clearly not understood yet how this stuff works.